crates/re_query_cache/src/range.rs

```diff
@@ -0,0 +1,147 @@
+        #[doc = "(combined) for `" $N "` point-of-view components and `" $M "` optional components."]
+        #[allow(non_snake_case)]
+        pub fn [<query_archetype_range_pov$N _comp$M>]<'a, A, $($pov,)+ $($comp,)* F>(
+            store: &'a DataStore,
+            query: &RangeQuery,
+            entity_path: &'a EntityPath,
+            mut f: F,
+        ) -> ::re_query::Result<()>
+        where
+            A: Archetype + 'a,
+            $($pov: Component + Send + Sync + 'static,)+
+            $($comp: Component + Send + Sync + 'static,)*
+            F: FnMut(
+                (
+                    (Option<TimeInt>, RowId),
+                    MaybeCachedComponentData<'_, InstanceKey>,
+                    $(MaybeCachedComponentData<'_, $pov>,)+
+                    $(MaybeCachedComponentData<'_, Option<$comp>>,)*
+                ),
+            ),
+        {
+            let mut iter_results = |bucket: &crate::CacheBucket| -> crate::Result<()> {
+                re_tracing::profile_scope!("iter");
+
+                let it = itertools::izip!(
+                    bucket.iter_data_times(),
+                    bucket.iter_pov_instance_keys(),
+                    $(bucket.iter_component::<$pov>()
+                        .ok_or_else(|| re_query::ComponentNotFoundError(<$pov>::name()))?,)+
+                    $(bucket.iter_component_opt::<$comp>()
+                        .ok_or_else(|| re_query::ComponentNotFoundError(<$comp>::name()))?,)*
+                ).map(|((time, row_id), instance_keys, $($pov,)+ $($comp,)*)| {
+                    (
+                        (Some(*time), *row_id), // TODO(cmc): timeless
+                        MaybeCachedComponentData::Cached(instance_keys),
+                        $(MaybeCachedComponentData::Cached($pov),)+
+                        $(MaybeCachedComponentData::Cached($comp),)*
+                    )
+                });
+
+                for data in it {
+                    f(data);
+                }
+
+                Ok(())
+            };
+
+            fn upsert_results<'a, A, $($pov,)+ $($comp,)*>(
+                arch_views: impl Iterator<Item = (Option<TimeInt>, re_query::ArchetypeView<A>)>,
+                bucket: &mut crate::CacheBucket,
+            ) -> crate::Result<u64>
+            where
+                A: Archetype + 'a,
+                $($pov: Component + Send + Sync + 'static,)+
+                $($comp: Component + Send + Sync + 'static,)*
+            {
+                re_log::trace!("fill");
+
+                let now = web_time::Instant::now();
+
+                let mut added_entries = 0u64;
+                let mut added_size_bytes = 0u64;
+
+                for (data_time, arch_view) in arch_views {
+                    let data_time = data_time.unwrap_or(TimeInt::MIN); // TODO(cmc): timeless
+
+                    // TODO(cmc): can't be relying on the primary RowId here, we need the max RowId.
```

Comment:
No, actually that's correct. What's incorrect is the mix of latest-at and range semantics in the same cache line, but that gets fixed in a follow-up PR.